src/main/java/org/magicdgs/http/jsr203/HttpPath.java

```diff
@@ -148,6 +250,148 @@ public WatchKey register(final WatchService watcher, final WatchEvent.Kind<?>...
+        return h;
+    }
+
+    @Override
+    public String toString() {
+        // TODO - should cache lazily?
+        final StringBuilder sb = new StringBuilder(authority);
+        sb.append(new String(normalizedPath, HttpUtils.HTTP_PATH_CHARSET));
+        return sb.toString();
+    }
+
+
+    /////////////////////////////////////////
+    // helper method for path as byte[]
+
+    private static byte[] getNormalizedPathBytes(final String path) {
+        // TODO - remove when relative HTTP/S path are supported
+        if (!path.isEmpty() && !path.startsWith(HttpUtils.HTTP_PATH_SEPARATOR_STRING)) {
+            throw new IllegalArgumentException("Relative HTTP/S path are not supported");
+        }
+
+        if (HttpUtils.HTTP_PATH_SEPARATOR_STRING.equals(path) || path.isEmpty()) {
+            return new byte[0];
+        }
+        final int len = path.length();
+
+        char prevChar = 0;
+        for(int i = 0; i < len; i++) {
+            char c = path.charAt(i);
+            if (isDoubleSeparator(prevChar, c)) {
+                return getNormalizedPathBytes(path, len, i - 1);
+            }
+            prevChar = checkNotNull(path, c);
+        }
+        if (prevChar == HttpUtils.HTTP_PATH_SEPARATOR_CHAR) {
+            return getNormalizedPathBytes(path, len, len - 1);
+        }
+
+        return path.getBytes(HttpUtils.HTTP_PATH_CHARSET);
+    }
+
+    private static byte[] getNormalizedPathBytes(final String path, final int len, final int offset) {
+        // get first the last offset
+        int lastOffset = len;
+        while (lastOffset > 0
+                && path.charAt(lastOffset - 1) == HttpUtils.HTTP_PATH_SEPARATOR_CHAR) {
+            lastOffset--;
+        }
+        if (lastOffset == 0) {
+            // early termination
+            return new byte[] {HttpUtils.HTTP_PATH_SEPARATOR_CHAR};
+        }
+        // byte output stream
+        try (final ByteArrayOutputStream os = new ByteArrayOutputStream(len)) {
+            if (offset > 0) {
+                os.write(path.substring(0, offset).getBytes(HttpUtils.HTTP_PATH_CHARSET));
+            }
+            char prevChar = 0;
+            for (int i = offset; i < len; i++) {
+                char c = path.charAt(i);
+                if (isDoubleSeparator(prevChar, c)) {
+                    continue;
+                }
+                prevChar = checkNotNull(path, c);
+                os.write(c);
+            }
+
+            return os.toByteArray();
+        } catch (IOException e) {
+            throw new RuntimeException("Should not happen");
+        }
+    }
+
+    private static boolean isDoubleSeparator(final char prevChar, final char c) {
+        return c == HttpUtils.HTTP_PATH_SEPARATOR_CHAR
+                && prevChar == HttpUtils.HTTP_PATH_SEPARATOR_CHAR;
+    }
+
+    private static char checkNotNull(String input, char c) {
+        if (c == '\u0000') {
+            throw new InvalidPathException(input, "Nul character not allowed");
```

Comment:
Typo: `Nul` -> `Null`

In addition, this should throw `IllegalArgumentException` instead; the `InvalidPathException` should be thrown in the FS, and might be catching this exceptions. This shouldn't happen in other places while normalizing, etc...

Reply:
Done. And removed the `String input` (not necessary).